Refactor hue to split bridge support from light platform

andreacampi opened the pull request:
## Description:

This is in preparation to supporting sensors.

**Related issue (if applicable):** fixes #10609

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4099

## Example entry for `configuration.yaml` (if applicable):
```yaml
hue:
  bridges:
    - host: 192.168.0.40
    - host: 192.168.0.80
    - filename: /tmp/phue.conf
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

andreacampi committed:
Introduce a new Hue component that knows how to talk to a Hue bridge, but doesn't actually set up lights.

andreacampi committed:
Refactor the hue lights platform to use the HueBridge class from the hue component.

homeassistant commented:
Hi @andreacampi, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#10691).

Once you do that we will be able to review and accept this pull request.

Thanks!

andreacampi commented:
There are a few things I still have to clean up, but I wanted to get the ball rolling and get early feedback.

To do:
* support multiple switches;
* support the older config schema and provide upgrade instructions;
* add back support for discovery.

Re config, this changes from:

```
light:
  - platform: hue
    host: DEVICE_IP_ADDRESS
```

to:

```
hue:
  - host: DEVICE_IP_ADDRESS
```

I find this makes more sense when the component can support more than just lights, yes?

andreacampi commented:
cc @robmarkcole

homeassistant commented:
Hi @andreacampi, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#10691).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @andreacampi, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#10691).

Once you do that we will be able to review and accept this pull request.

Thanks!

balloob commented:
It seems like you have some commits in this branch that are not yours, please clean base your work off a fresh copy of the dev branch.

balloob commented:
I haven't looked at the code at all, but yes, migrating hue from just a platform to be a component for auth and a platform for lights = 👍

robmarkcole commented:
Im not sure I follow this refactor, appears to be a lot of light specific code in `components/hue.py ` which is the hub/auth..?

andreacampi commented:
@robmarkcole yeah that's fair, not sure why I decided to go that way. I'll clean that up and/or explain.

andreacampi commented:
I'll rebase and check the failed tests; just not today as I've had a long day on ~4 hours of sleep.

andreacampi committed:
Reimplement support for multiple bridges

robmarkcole commented:
@andreacampi looks good and new config looks suitable. Cheers!

andreacampi committed:
Auto discover bridges.

balloob commented on homeassistant/components/hue.py:
```diff
@@ -0,0 +1,245 @@
+    for bridge in bridges:
+        filename = bridge.get(CONF_FILENAME, PHUE_CONFIG_FILE)
+        allow_unreachable = bridge.get(CONF_ALLOW_UNREACHABLE,
+                                       DEFAULT_ALLOW_UNREACHABLE)
```
Please set defaults in the `BRIDGE_CONFIG_SCHEMA`, pass in `default=DEFAULT_ALLOW_UNREACHABLE` into `vol.Optional`

balloob commented on homeassistant/components/hue.py:
```diff
@@ -0,0 +1,245 @@
+        allow_hue_groups = bridge.get(CONF_ALLOW_HUE_GROUPS,
+                                      DEFAULT_ALLOW_HUE_GROUPS)
+
+        host = bridge.get(CONF_HOST, None)
```
`None` is the default for `get`.

balloob commented on homeassistant/components/hue.py:
```diff
@@ -0,0 +1,245 @@
+
+    try:
+        with open(path) as inp:
+            return next(json.loads(''.join(inp)).keys().__iter__())
```
jeeeez.

- Use `json.load(inp)` instead of converting file pointer to string.
- Call `iter(obj)` instead of `obj.__iter__()`

balloob commented on homeassistant/components/light/hue.py:
```diff
@@ -60,251 +46,142 @@
+    setup_data(hass)
 
+    for bridge in hass.data[hue.DOMAIN].values():
+        update_lights(hass, bridge, add_devices)
```
This cannot work because you throttle `update_lights`, so only the first call will go through.

balloob commented on homeassistant/components/hue.py:
```diff
@@ -0,0 +1,245 @@
+    filename = 'phue-{}.conf'.format(serial)
+    setup_bridge(host, hass, filename)
+
+    discovery.load_platform(hass, 'light', DOMAIN, discovery_info, {})
```
We should fire this inside `HueBridge.setup()` and pass the hue bridge ID as discovery info. The light platform can then pick the right bridge and just expose those entities.

Light platform should also make sure to cancel setup when `discovery_info is None`, as a discovery event is always fired when a bridge is connected.

balloob commented on homeassistant/components/light/hue.py:
```diff
@@ -20,31 +18,19 @@
-DEFAULT_ALLOW_UNREACHABLE = False
-DOMAIN = "light"
-SERVICE_HUE_SCENE = "hue_activate_scene"
+DOMAIN = "hue.lights"
```
Let's not call it `DOMAIN` as it is not its DOMAIN. Instead use `DATA_KEY`

balloob commented on homeassistant/components/light/hue.py:
```diff
@@ -60,251 +46,142 @@
+        new_lightgroups = process_groups(
+            hass, api, bridge, bridge_type,
+            lambda **kw: update_lights(hass, bridge, add_devices, **kw))
+        new_lights = new_lights + new_lightgroups
```
```python
new_lights.extend(new_lightgroups)
```

balloob commented on homeassistant/components/light/hue.py:
```diff
@@ -60,251 +46,142 @@
-        scene_name = call.data[ATTR_SCENE_NAME]
-        bridge.run_scene(group_name, scene_name)
+    for light_id, info in api_lights.items():
+        if light_id not in hass.data[DOMAIN]['lights']:
```
To avoid extra key lookups, store the lights as local variable

```python
lights = hass.data[DOMAIN]['lights']
```

balloob commented on homeassistant/components/light/hue.py:
```diff
@@ -60,251 +46,142 @@
+def setup_data(hass):
+    """Initialize internal data. Useful from tests."""
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {'lights': {}, 'lightgroups': {}}
```
let's replace `lights` and `lightgroups` with `DATA_LIGHTS` and `DATA_LIGHTGROUPS` constants

balloob commented on homeassistant/components/light/hue.py:
```diff
@@ -60,251 +46,142 @@
-        """Set up actions to do when our configuration callback is called."""
-        setup_bridge(host, hass, add_devices, filename, allow_unreachable,
-                     allow_in_emulated_hue, allow_hue_groups)
+        if lightgroup_id not in hass.data[DOMAIN]['lightgroups']:
```
idem:

```python
groups = hass.data[DOMAIN]['lightgroups']
```

balloob reviewed (changes requested):
This is great work! Just a few minor changes and this will be good to go 🥇 .

andreacampi commented:
Thanks for the great feedback. I'm pretty happy that most of it is existing code I was too lazy to fix in the refactoring, but happy to do it now :D I'll get on it.

andreacampi committed:
Provide some migration support by showing a persistent notification.

andreacampi committed:
Address most feedback from code review.

andreacampi committed:
Call load_platform from inside HueBridge.setup passing the bridge id.

Not only this looks nicer, but it also nicely solves additional bridges being added after initial setup (e.g. pairing a second bridge should work now, I believe it required a restart before).

andreacampi commented:
OK, latest push addressed all concerns from @balloob. Ready for one more round of feedback :)

andreacampi committed:
Add a unit test for hue_activate_scene

andreacampi commented:
I noticed some of the tests were red on 3.5, fixed that. Not sure why coveralls is complaining give the coverage increased; and also it's mostly concerned with a file (homeassistant/components/discovery.py) I barely touched. I'd ignore that.

andrey-git commented on .coveragerc:
```diff
@@ -85,6 +85,9 @@ omit =
     homeassistant/components/hive.py
     homeassistant/components/*/hive.py
 
+    homeassistant/components/hue.py
```
You are adding tests (thank you!) so no need to add the files to .coveragerc

andreacampi replied:
So should I drop all the changes to .coveragerc or only this line (`homeassistant/components/hue.py`)?
`homeassistant/components/light/hue.py` was already in the file some way down, should I live it there?

andrey-git replied:
.coveragerc includes file not covered by tests, so remove both homeassistant/components/hue.py and homeassistant/components/*/hue.py

andrey-git commented:
Is it possible to write the tests to work with a mock of the library  instead of importing it?
cc @balloob for his opinion on this.

MartinHjelmare commented on homeassistant/components/light/hue.py:
```diff
@@ -75,236 +72,170 @@
-        host = discovery_info.get('host')
-    else:
-        host = config.get(CONF_HOST, None)
+    return True
```
Remove this.

MartinHjelmare commented on homeassistant/components/light/hue.py:
```diff
@@ -1,50 +1,43 @@
-DEFAULT_ALLOW_UNREACHABLE = False
-DOMAIN = "light"
-SERVICE_HUE_SCENE = "hue_activate_scene"
+DATA_KEY = 'hue.lights'
```
`'hue_lights'`

andreacampi commented on homeassistant/components/hue.py:
```diff
@@ -0,0 +1,245 @@
+    bridges = config.get(CONF_BRIDGES, [])
+    if len(bridges) == 0:
+        _LOGGER.info("No bridges found in configuration")
+        return True
```
Yeah, I realized that but I thought it might be nice to print something. But I'm not sure it's really useful--after playing with this I'd just recommend letting discovery take care of things.

andreacampi committed:
Address feedback from code review.

andreacampi committed:
After feedback from @andrey-git I was able to find a way to not import phue in tests, yay!

andreacampi commented:
So, question: should I squash all commits together? I like working with small commits but this is getting a bit messy :)

MartinHjelmare commented:
Please don't squash until code review is finished. Squashing makes it hard to follow changes. We always squash when we merge, so the contributor is never required to squash.

andreacampi commented:
Hey folks, how do we move forward with this?

andrey-git commented on script/gen_requirements_all.py:
```diff
@@ -58,6 +58,7 @@
     'numpy',
     'paho-mqtt',
     'pexpect',
+    'phue',
```
Since now you mock phue in tests - remove this line.

To make sure this works - uninstall phue from your dev enviroment and check that the tests still pass.

andreacampi committed:
Inject a mock phue in a couple of places

andreacampi commented:
This is an unrelated test that failed. I can try rebasing I guess.

balloob commented:
🎉  🐬  Awesome, great job !

mezz64 commented:
These changes seem to be allowing HASS to discover it's own emulated hue bridge as a bridge device and throws an error because it can't find the config file.

m2k2 commented:
I'm also having some problems since this commit after re-working my configuration. Once the hubs are added with emulated hue disabled, I see the error below for each light. Should I open up a new issue?

```
Update for light.garage__right fails
Traceback (most recent call last):
  File "/srv/homeassistant/homeassistant_venv/lib/python3.5/site-packages/homeassistant/helpers/entity.py", line 199, in async_update_ha_state
    yield from self.async_device_update()
  File "/srv/homeassistant/homeassistant_venv/lib/python3.5/site-packages/homeassistant/helpers/entity.py", line 306, in async_device_update
    yield from self.hass.async_add_job(self.update)
  File "/usr/lib/python3.5/asyncio/futures.py", line 380, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 304, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 293, in result
    raise self._exception
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/srv/homeassistant/homeassistant_venv/lib/python3.5/site-packages/homeassistant/components/light/hue.py", line 385, in update
    self.update_lights(no_throttle=True)
  File "/srv/homeassistant/homeassistant_venv/lib/python3.5/site-packages/homeassistant/components/light/hue.py", line 149, in <lambda>
    lambda **kw: update_lights(hass, bridge, add_devices, **kw))
  File "/srv/homeassistant/homeassistant_venv/lib/python3.5/site-packages/homeassistant/util/__init__.py", line 306, in wrapper
    result = method(*args, **kwargs)
  File "/srv/homeassistant/homeassistant_venv/lib/python3.5/site-packages/homeassistant/util/__init__.py", line 306, in wrapper
    result = method(*args, **kwargs)
TypeError: update_lights() got an unexpected keyword argument 'no_throttle'
```

andreacampi commented:
Both of these are pretty bad :( I'll try to report and open tasks as needed, but @mezz64 @m2k2 if you want to do that'd be helpful.

@m2k2 unless I broke something last second while iterating on unit tests, I don't know that I have enough to reproduce it. I'll ping you if I have trouble.

andreacampi commented:
@balloob thoughts on me removing the `@Throttle` annotation? that would remove the issue @m2k2 is seeing.

andreacampi commented:
See https://github.com/home-assistant/home-assistant/pull/11126

andreacampi commented:
Another pull request: https://github.com/home-assistant/home-assistant/pull/11128

brent20 commented:
How would one set the scan_interval now on the new hue component? Before I was able to set it per the docs here: `https://home-assistant.io/docs/configuration/platform_options/`

Leaving this in place with the new Hue component seemed to break the configuration. Once I know the new way to set a scan_interval, I'll go ahead and submit a PR to update the docs. 

Thanks!

MartinHjelmare commented:
Please open an issue if you suspect a bug. If you need help please use our help channels:
https://home-assistant.io/help/#communication-channels

Merged PRs should not be used for support or bug reports. Thanks!

mcoms commented:
This work appears to [change the filename](https://github.com/home-assistant/home-assistant/pull/10691/files#diff-8c8be3774802c497e688748abb9c9c50R113) of the default `phue.conf` to `phue-<bridge serial number>.conf` for discovered bridges. This causes HomeAssistant to boot with:

```
2017-12-28 10:15:26 INFO (MainThread) [homeassistant.setup] Setting up hue
2017-12-28 10:15:26 INFO (MainThread) [homeassistant.setup] Setup of domain hue took 0.0 seconds.
2017-12-28 10:15:26 INFO (MainThread) [homeassistant.core] Bus:Handling <Event component_loaded[L]: component=hue>
2017-12-28 10:15:26 INFO (MainThread) [homeassistant.core] Bus:Handling <Event platform_discovered[L]: service=philips_hue, discovered=host=192.168.2.157, port=80, ssdp_description=http://192.168.1.150:80/description.xml, name=Fhue (192.168.1.150), model_name=Philips hue bridge 2015, model_number=BSB002, serial=00178848ac26>
2017-12-28 10:15:26 INFO (SyncWorker_6) [phue] Attempting to connect to the bridge...
2017-12-28 10:15:26 INFO (SyncWorker_6) [phue] Error opening config file, will attempt bridge registration
2017-12-28 10:15:26 WARNING (SyncWorker_6) [homeassistant.components.hue] Connected to Hue at 192.168.1.150 but not registered.
2017-12-28 10:15:27 INFO (SyncWorker_6) [homeassistant.loader] Loaded configurator from homeassistant.components.configurator
2017-12-28 10:15:27 INFO (MainThread) [homeassistant.core] Bus:Handling <Event service_registered[L]: domain=configurator, service=configure>
2017-12-28 10:15:27 INFO (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=configurator.philips_hue, old_state=None, new_state=<state configurator.philips_hue=configure; configure_id=139857224447368-1, fields=[], friendly_name=Philips Hue, entity_picture=/static/images/logo_philips_hue.png, description=
Press the button on the bridge to register Philips Hue with Home Assistant.

![Location of button on bridge](/static/images/config_philips_hue.jpg)
, submit_caption=I have pressed the button @ 2017-12-28T10:15:27.009680+00:00>>
```

And to fail to connect to the bridge.

My hue bridge is in the roof and pressing the button was quite a pain.

It would be handy to either, upgrade (i.e. rename) the existing `phue.conf` file automatically, or, to have a log message showing the filename the component was trying to read, so I could manually rename it.

MartinHjelmare commented:
Please open an issue in the appropriate reoo if you suspect a bug or there's a problem with documentation. If you need help please use our help channels:
https://home-assistant.io/help/#communication-channels

Merged PRs should not be used for support or bug reports. You can link to the PR that you want to tag from the issue. Thanks!
